### DISCUSSION TITLE
spatial-omics workflow specs

### DISCUSSION THREAD

**Discussion by philipmac** [niaid/image_portal_workflows#227] [General]
Terms:
Nanostring - a device/instrument used to generate data for this workflow.
DSP - instrument
ROI region of interest

Pipepine 1: "Large 2d color images pipeline".
1) large pngs
2) large czi

1) IFF czi file, convert to tiff.
2) convert from tiff || png -> pyramidal choice.

Pipeline 2: "multi channel pipeline (2d)"
Typically light microscopy / Florescent. Possibly turn channels on off
Computationally similar to above, but will have ability to do things specific to a channel.
(Typically for now - we'll use czi.)

 ->
saved as png file -> very large (eg 3 gig), want to make this available to viewer. 
Additionally this is saved as a CZI ("raw") file (~order of mag larger). Want to make this available to the viewer too.

Outputs:
pyramid format, viewable in a web viewer. (Maybe a thumb.)

**Comment by hansenbry** [comment]
Don't know if it helps or matters for the notes, but DSP = Digital Spatial Profiler

Also the 3GB png is an export of a "raw" tiff file separate from the CZI workflows

**Comment by philipmac** [comment]
Thanks @hansenbry  - to us there doesn't seem to be one single, czi workflow. 

There seems to be two workflows, one which deals with large color images (which may or may not take CZI type inputs) - but if we DO take a czi input we do NOT care about having special treatments for the channels, we'll transform it as a color image into some zoomable format.

The other workflow, involved with florescence - (which will take czi inputs for now), we DO care about the ability to do certain treatments with certain channels. 

Thoughts welcome.

**Comment by blowekamp** [comment]
The CZI files ( both H&E and FL ) will contain multiple images.  Including a "label image" and a "macro image". These images are smaller and may not need a multi-scale viewer. Perhaps just a JPEG/PNG conversion? There will also be multiple "main" images and lower resolutions of them for a pyramid like format. These "main" images will need to be converted to our web based multi-scale format.

Additionally, I noted that when viewing RGB images in the Zen viewer it appeared to apply to a gama of .45 to the RGB channels.

**Comment by philipmac** [comment]
May 22 2023
Regarding file type in <SPREADSHEET>

What relevance have these files?
Associated with experiment / project in some form. Will want to associate them with experiment in some way. Will be "clickable" (and available for download).

All "small" files, ie: XLS / PDF / ncc / pks files should be visible to users, and available for download as a single package per collection. (Maybe - don't know if we care or not about this.)

A "sample" (AKA a "collection") is a set of files in this spreadsheet. Once this is noticed by Hedwig it can be submitted by user to pipeline.

The only files that the pipeline currently cares about are: the _n_ large png file, and _n_ czi files. (if n > 50 then throw error and don't run.)


Outputs - generate thumb from "large" png file. 
Heatmap "thumb" is too low resolution.

**Reply by NetaFG** [reply]
Will want to associate them with experiment in some way ->We will want to associate
 them with a "session"
The only files that the pipeline currently cares about are: the n large png file, and n czi files. (if n > 50 then throw error and don't run.) and a "small" png file.


<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/FILIPG~1/AppData/Local/Temp/1/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/FILIPG~1/AppData/Local/Temp/1/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



File | Current Source | File Type | NUMBER OF FILES | Est. Size | Visualized? | Level | FOLDER
-- | -- | -- | -- | -- | -- | -- | --
ROI Overlays | DSP Export | .png/other | UP TO 8 | 2-5 Gb | Yes ( Neuroglancer) | Session | "COLLECTION"="SAMPLE"
Heatmap RAW | DSP Export | .png | 1 | <1 Mb | Yes (carousel) | Session | "COLLECTION"="SAMPLE"
Heatmap QC | DSP Export | .png | 1 | <1 Mb | Yes (carousel) | Session | "COLLECTION"="SAMPLE"
H&E Images | Axio Scan   Z1/Investigator | .czi/other | UP TO 8 | ~25 Gb | Yes - Neuroglancer   (Add ROI areas) | Session | "COLLECTION"="SAMPLE"
IF Images | Axio Scan Z1/Other   scope | .czi/other | UP TO 8 | ~50 Gb | Yes - Neuroglancer (4   channel) | Session | "COLLECTION"="SAMPLE"



</body>

</html>
Please note that the pipeline is required to identify the images to be processed and generate adequate assets for each image type.

**Comment by blowekamp** [comment]
I checked a couple "ROI Overlay" png files of exif meta data. The tools I used did not give and additional types beyond the ones to describe the image data e.g. pixel type, resolution etc.

Also for reference the "heat map" image I inspected was 2.6x2.9k size in pixels.

**Reply by philipmac** [reply]
OK - looks like next best option would to ask that every png file to be processed have some substr prepended to file name.
This would allow me to ask for: 
``` <substr>*.png```

**Reply by NetaFG** [reply]
Bryan and Dan are not familiar with the substr function:
Here is what Dan offered:
1) 
If I perform the import using the new Nanostring function, I think it will double the number of images. First set I can append as "*_Scan" which I can run in "scan only" mode. Ideally these would be uploaded to Hedwig, worked on by investigators for ROI selection in Neuroglancer then saved. I would then download the images and import them back into the DSP to use the overlay. I would make collections, then export the final overlaid ROI images back into Hedwig. I could append this second set as "*_ROI". I should be able to see how the first half of this works tomorrow. 

2)
I suppose the other approach is to always make the smaller file "Heatmap" which I think is the default and would save me from having to append so many image files names.

**Comment by blowekamp** [comment]
A "brainstorming" idea for the heat maps which would require more work. Instead of being given a "png" for the heat map figure. The input data could be given and and interactive web visualization could be generated with python or R shiny. It could be implemented such that when a block is hovered over a pop up gives the details.

https://jokergoo.github.io/InteractiveComplexHeatmap/articles/shiny_dev.html

**Reply by NetaFG** [reply]
This tool is so cool!

**Comment by philipmac** [comment]
@hansenbry - Looking at the files that have been presented as ROI, we notice that they are all tiffs (not pngs, as suggested in the table above).
eg 
```
$ ls /mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/*/ROI/*
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/129 Female Naive.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/129 Male 3 DPI.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/129 Male 6 DPI.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/129 Male Naive.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/A_J Female 3 DPI.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/A_J Female 6 DPI.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/A_J Female Naive.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/A_J Male 3 DPI.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/A_J Male 6 DPI.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/A_J Male Naive.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/B6J Female 3 DPI.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/B6J Female 6 DPI.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/B6J Female Naive.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/B6J Male 3 DPI.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/B6J Male 6 DPI.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/B6J Male Naive.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/Balb Female 3 DPI.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/Balb Female 6 DPI.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/BALB female Naive.tiff
/mnt/ai-fas12/RMLEMHedwigQA/Projects/RTB/efischer/hansenbry-2023-0329-Spatial/Light-Microscopy-2023-0403-Option1/ROI/Balb Male 3 DPI.tiff

```

We note that the version of GraphicsMagick is 1.3.38 2022-03-26 is currently being used. 
We've requested Imagemagick  be installed. 

For `gm` command:
Suspect `sharpen` is not helpful.
Suspect `quality` should be pushed up to 90.

For Imagemagick:
Suspect `clahe` argument might be useful.

**Reply by blowekamp** [reply]
The about TIFF files are considered old, and not representative of the currently expected PNG files.

**Comment by philipmac** [comment]
Overlay png (ROI) - do not sharpen.

**Comment by philipmac** [comment]
Note - we'll want to create thumbs - same size as other thumbs (?) - for the heatmaps.

**Comment by blowekamp** [comment]
The steps for processing the "large png" images or ROI images consist the following steps:

0. Convert the PNG to TIFF
1. Convert the TIFF o a OME-NGFF ZARR format.
2. Update the ZARR structure with chucking required for Neuroglancer. 
3. Create a 2d thumbnail image from the ZARR structure.
4. Create a 2d key image from the ZARR structure.

**Reply by blowekamp** [reply]
Step 1: Generate OMG-NGFF ZARR format

The [ bioformats2raw ](https://github.com/glencoesoftware/bioformats2raw) needs to be used to convert the large png to the OME-
NGFF zarr pyramid format supported by Neuroglancer. 

The input png files may be RGBA ( 4-channel ) with the A ( alpha ) not containing useful information.

The following is the recommended command line for the large PNG files:

```
bioformats2raw --scale-format-string '%2$d' --downsample-type AREA --compression=blosc --compression-properties cname=zstd --compression-properties clevel=5 --compression-properties shuffle=1  roi.png output.zarr
```

The compression properties were selected based to the reported performance of images of noise from [this paper.](https://www.biorxiv.org/content/biorxiv/early/2023/01/25/2023.01.24.525380.full.pdf) . This choice follows the recommendation of the conclusion.

**Reply by blowekamp** [reply]
Step 2:

Re-chunk  the ZARR structure so that the RGB(A) channels are not split between chunks with the `zarr_rechunk` command provided by tomojs-pytools. The following command line should be run:

`zarr_rechunk --chunk-size 512 output.zarr`

The ZARR structure is updated in place so that the chunks are compatible for display in Neuroglancer. The resulting chunks are 780KB to 1MB size is the current estimate for an appropriate chunk size used for visualization in Neuroglancer.

Note: The Alpha channel could also be removed in this step as it does not contain information ( it's all zeros ), and would reduce storage, bandwidth, and computational time.

**Reply by blowekamp** [reply]
Step 3: Create a thumbnail image.

A thumbnail image can be created by extracting a level from the ZARR pyramid structure and resizing it to the desired resolution, then saving. The following Python code does this:


```
zarr_array_path = "output.zarr"
thumbnail_size = [300,300]
jpeg_quality = 90

sitk_image = pytools.zarr_extract_2d(zarr_array_path, target_size_x=thumbnail_size[0], target_size_y=thumbnail_size[1])
SimpleITK.WriteImage(sitk_image, "output.jpeg", useCompression=True, compressionLevel=jpeg_quality)
```

There are a couple filtering steps done in the generations of the thumbnails to prevent aliasing. The initial plan is not to perform and sharping. And additional SimpleITK filter could be run on `sitk_image` to sharpen the image if later determined a requirement.

Step 4: Generate a key image

The same as above except that the size will be different.

**Reply by blowekamp** [reply]
Step 0: Convert the input PNG to TIFF

The input png uses some form of compression that slows down bioformats2raw ~100X. Converting to a uncompressed TIFF dramatically speeds up the process.

The unused alpha channel can also be removed. (#249 ) The following command can be used to perform the conversion with imagemagick:

```
convert input.png -background white -alpha remove -alpha off -compress temp.tiff
```

**Comment by blowekamp** [comment]
The following are the step to process large CZI files of immunofluorescence (IF) microscopy. This file consists of multiple IF images ( at multiple resolutions ) and also a "label" and "macro" color/RGB images. The IF images are multi-channel ( not visual spectrum color/RGB ) and are generally visualized with false color applied to each channel. 

The IF images are visualized with Neuroglancer, and no false colored thumbnails or key image, but do require neuroglacer metadata for visualization.

The "label" and "macro" color/RGB images need a thumbnail and full resolution image ( no key image, as the specs for a key image are greater resolution than the original source image ).

Overview
Step A: convert CZI file to OME-NGFF ZARR format with OME XML.
Step B:  Generate neroglancer meta-data from zarr sub-image for each IF image.
Step C: Create thumbnail image from zarr sub-image label ( maybe macro )

**Reply by blowekamp** [reply]
Step A: Runs ones on the original CZI file.

Step A.1: Convert CZI to ZARR.
```
 bioformats2raw -p --max_workers=$nproc --downsample-type AREA --compression=blosc --compression-properties cname=zstd --compression-properties clevel=5 --compression-properties shuffle=1  IF.czi output_IF.zarr
```

Note the performance of the above conversion scale very will with the number of workers, so this is an important parameter to get reasonable performance.

Step A.2: Rechunk the ZARR arrays.
Re-chunk the ZARR structure so that multi-channel/RGB channels are not split between chunks with the zarr_rechunk command provided by tomojs-pytools. The following command line should be run:

```
zarr_rechunk --chunk-size 512 output.zarr
```

The ZARR structure is updated in place so that the chunks are compatible for display in Neuroglancer. This will rechunk all the extracted images in the ZARR structured.


The above commands converts all images in the CZI file into one ZARR structure. In Zarr terminology the root of `output_IF.zarr` is a group containing each OME-NGFF image ( another group ). Practically, `output_IF.zarr` is a directory with subdirectories 0-N of the OME-NGFF images. A directory such as `output_IF.zarr/0` is a valid OME-NGFF image and can be passed that way to application such as neruoglancer.

**Reply by blowekamp** [reply]
Step B: Runs for each IF image in the output ZARR structure.

Step B.1:

Determine which ZARR sub-images are IF and multi-channel. TODO

Step B.2:  Generate meta-data for multi-channel visualization in Neuroglancer.

Run a pytools functions called `generate_metadata_multichannel` on each multichannel subdirectory.
TODO this needs to be implemented.

**Reply by blowekamp** [reply]
Step C: Generate thumbnails for both the "label" and "macro" image. These images are RGB color and can be converted to standard web image file formats such as png or jpeg.

Step C.1: On label image detect text orientation and rotate the image.

Does the text orientation need to be detected? This would be a computational intensive task to detect the text. There are large existing libraries which do this but is this required?

If we can assume that we need a 90 degree clockwise rotation of an image, then the following SimpleITK code can be used:

```
rotated_img = sitk.Flip(sitk.PermuteAxes(img, [1,0]), [True,False])
```

Step C.2: Generate thumbnail
A thumbnail image can be created by extracting a level from the ZARR sub-image pyramid structure and resizing it to the desired resolution, then saving. The following Python code does this:

```
zarr_array_path = "output.zarr/1"
thumbnail_size = [300,300]
jpeg_quality = 90

sitk_image = pytools.zarr_extract_2d(zarr_array_path, target_size_x=thumbnail_size[0], target_size_y=thumbnail_size[1], auto_uint8=True)
# rotate 90 degrees clockwise if needed?
rotated_img = sitk.Flip(sitk.PermuteAxes(sitk_image, [1,0]), [True,False])
SimpleITK.WriteImage(rotated_img, "output.jpeg", useCompression=True, compressionLevel=jpeg_quality)
```


Step C.3: Generate full resolutions image

The same as above except that the size should be `[None, None]` to extract the full resolutions from the ZARR.
